torchvision/prototype/datapoints/_dataset_wrapper.py

```diff
@@ -0,0 +1,411 @@
+        return {}
+
+    dict_of_lists = {key: [value] for key, value in list_of_dicts[0].items()}
+    for dct in list_of_dicts[1:]:
+        for key, value in dct.items():
+            dict_of_lists[key].append(value)
+    return dict_of_lists
+
+
+def wrap_target_by_type(dataset, target, type_wrappers, *, fail_on=(), attr_name="target_type"):
+    if target is None:
+        return None
+
+    target_types = getattr(dataset, attr_name)
+
+    if any(target_type in fail_on for target_type in target_types):
+        raise RuntimeError(
+            f"{type(dataset).__name__} with target type(s) {sequence_to_str(fail_on, separate_last='or ')} "
+            f"is currently not supported by this wrapper. "
+            f"If this would be helpful for you, please open an issue at https://github.com/pytorch/vision/issues."
+        )
+
+    if not isinstance(target, (tuple, list)):
+        target = [target]
+
+    wrapped_target = tuple(
+        type_wrappers.get(target_type, identity)(item) for target_type, item in zip(target_types, target)
+    )
+
+    if len(wrapped_target) == 1:
+        wrapped_target = wrapped_target[0]
+
+    return wrapped_target
+
+
+@cache
+def get_categories(dataset):
+    categories_fn = {
+        datasets.Caltech256: lambda dataset: [name.rsplit(".", 1)[1] for name in dataset.categories],
+        datasets.CIFAR10: lambda dataset: dataset.classes,
+        datasets.CIFAR100: lambda dataset: dataset.classes,
+        datasets.FashionMNIST: lambda dataset: dataset.classes,
+        datasets.ImageNet: lambda dataset: [", ".join(names) for names in dataset.classes],
+        datasets.DatasetFolder: lambda dataset: dataset.classes,
+        datasets.ImageFolder: lambda dataset: dataset.classes,
+    }.get(type(dataset))
+    return categories_fn(dataset) if categories_fn is not None else None
+
+
+def classification_wrapper(dataset, sample):
+    image, label = sample
+    if label is not None:
+        label = datapoints.Label(label, categories=get_categories(dataset))
+    return image, label
+
+
+for dataset_type in [
+    datasets.Caltech256,
+    datasets.CIFAR10,
+    datasets.CIFAR100,
+    datasets.ImageNet,
+    datasets.MNIST,
+    datasets.FashionMNIST,
+    datasets.GTSRB,
+    datasets.DatasetFolder,
+    datasets.ImageFolder,
+]:
+    _WRAPPERS[dataset_type] = classification_wrapper
+
+
+def segmentation_wrapper(dataset, sample):
+    image, mask = sample
+    return image, datapoints.Mask(F.to_image_tensor(mask).squeeze(0))
+
+
+for dataset_type in [
+    datasets.VOCSegmentation,
+]:
+    _WRAPPERS[dataset_type] = segmentation_wrapper
+
+
+def video_classification_wrapper(dataset, sample):
+    if dataset.video_clips.output_format == "THWC":
+        raise RuntimeError(
+            f"{type(dataset).__name__} with output_format='THWC' is not supported by this wrapper, "
+            f"since it is not compatible with the transformations. Please use output_format='TCHW' instead."
+        )
```

Comment:
I've opted to fail here. Another option would be to just perform the permutation here, but that seems disruptive if the user explicitly set `output_format="THWC"` on the dataset. Plus, with that parameter they have very simple remedy if they want the default wrapping. 

Note that the error message is slightly different from the other ones: since our plan is to not support `"THWC"` in the transforms, there is no suggestion to open an issue if someone wants this.